### PR TITLE
Add scrolling to MissionPanel job/mission panes

### LINKED ISSUE #4174: Using keyboard keys to move in a long list of accepted missions should scroll the list to show the selected mission
**Describe the bug**
As a player with more than 50 current missions, I want the list of current missions to scroll to the mission highlighted on the list even when I am browsing the list with keyboard up and down keys, so that I can see the items around it.

**To Reproduce**
Case A:
1. Accept more missions that can be simultaneously shown on the list of current missions. (With my screen resolution, this is at least 50 missions.)
2. Go to the map screen ('m')
3. Click on the 'Missions' button. The first current mission on visible the list of current missions, but the last mission is not visible.
4. Select the first mission. The mission's details appear at the middle of the screen.
5. Press the up arrow key to move to the last mission on the list.
6. Last mission's details appear at the middle of the screen, but the selected mission is not visible on the list of current missions.

Alternatively (Case B):
5b. Press the down arrow key to move past the end of the visible part of the current missions list.

**Expected behavior**
When navigating to a non-visible item of a list using keyboard, I would expect the list to scroll to show the selected item. This should apply both for moving past the beginning (or end) of the list (Case A) and for moving past the visible end (or beginning) of the list (Case B).

**System (please complete the following information):**
 - OS: Windows 10 pro, 10.0.0.17763 Build 17763
 - Screen resolution: 1600x900, game running in non-fullscreen mode (window is maximized)
 - Game Source: https://github.com/endless-sky/endless-sky/releases
 - Version: 0.9.8

### PR DESCRIPTION
**Pull request by tehhowch**
Fixes #4174 via the short-term solution, which is to scroll to the selected mission upon click or keypress. The long-term solution is to use an interface definition (deferred for an unstable release as it is a sizable change).

~~Currently uses a "paginated" scroll, but I'm considering changing to a "smart" continuous scroll. I originally felt the paginated would provide a smoother feel but the large scrolls can feel disruptive.~~

Feedback welcomed.

Saves to test with:
 - [ManyMissions tester.txt](https://github.com/endless-sky/endless-sky/files/3366036/ManyMissions.tester.txt)
 - [ManyAvailable tester.txt](https://github.com/endless-sky/endless-sky/files/3390467/ManyAvailable.tester.txt)

Windows builds to test with:
 - [44755e9](https://ci.appveyor.com/project/tehhowch/endless-sky/builds/25804828/artifacts) (job list above map key, "paginated" scrolling)
 - [bbb4d29](https://ci.appveyor.com/project/tehhowch/endless-sky/builds/25971743/artifacts) (job list behind map key)
 - [e4aee12](https://ci.appveyor.com/project/tehhowch/endless-sky/builds/25972758/artifacts) ("smart" scrolling)

### CODE REVIEW

**Comment by ChamEV** [comment]
It works well with the mouse scroll wheel, but can lag behind if using the keyboard's up and down arrows.

The list window should not extend so far down that it's behind the shipyard navigation and map zoom buttons.

Consider adding icons like these when there is a need to scroll up or down, centered at the top/bottom of the window when appropriate.

<details>
<summary>Click to expand!</summary>

image
image
</details>

The double downward arrows without the circle would be my choice.

The arrow used to close/open categories on the shipyard/outfitter window would be a good choice as well.  images/ui/expanded.png

image

These could also be added to the shipyard/outfitter map lists on the left for consistency.

**Comment by tehhowch** [comment]
For this quick fix I have no intention of anything beyond scrolling. In the linked issue I mention the better fix is to overhaul the panel to use interface definitions, as then the panel layout can be changed by mods. That, however, is too large a change for a stable release.

With regards to held scroll keys, I'm fine with it lagging behind. I used the same scroll step magnitude as used in centering on the selected system, so they should stop scrolling in unison.

**Comment by Amazinite** [comment]
>The list window should not extend so far down that it's behind the shipyard navigation and map zoom buttons.

Notably, the offered missions list will also extend to be *in front* of the map key, instead of behind it like the active missions. 

image

As a temporary solution this works, but I do agree that at some point these lists of missions should be capped in size.

**Comment by tehhowch** [comment]
Switched to a different scroll style. @Amazinite @ChamEV @petervdmeer please share feedback.

**Comment by Amazinite** [comment]
Still looks good. I don't really have any preference for one or the other though.
Something I have noticed with both scrolling methods is that if I hold down my arrow key then I will select a different mission faster than the list can scroll, so the scrolling never catches up until I stop holding down the arrow key, at which point it quickly snaps to the selected mission. Preferably, holding down the arrow key would have the scrolling keep up with the selected mission.

File: source/MissionPanel.cpp [hunk 1]
```diff
@@ -64,10 +66,52 @@ namespace {
 		
 		return false;
 	}
+	
+	size_t MaxDisplayedMissions(bool onRight)
+	{
```
[the remark this answers is not in the document]
Comment (tehhowch) [reply source/MissionPanel.cpp]: Magic values will be getting cleaned up in the previously mentioned refactoring

File: source/MissionPanel.cpp [hunk 2]
```diff
@@ -64,10 +66,52 @@ namespace {
...
+		return static_cast<unsigned>(max(0, static_cast<int>(floor((Screen::Height() - (onRight ? 160. : 190.)) / 20.))));
```
[the remark this answers is not in the document]
Comment (tehhowch) [reply source/MissionPanel.cpp]: I had been going between ceil and floor while testing. This entire method will likely be removed in the refactoring.

File: source/MissionPanel.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/MissionPanel.cpp]: Leaving the magic values for refactoring is fine.

File: source/MissionPanel.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/MissionPanel.cpp]: As I said: not a big issue, keeping this for refactoring is also okay.

File: source/MissionPanel.cpp [hunk 3]
```diff
@@ -190,8 +253,10 @@ void MissionPanel::Draw()
 		AcceptedVisible());
 	DrawList(accepted, pos);
 	
+	// Now that the mission lists and map elements are drawn, draw the top-most UI elements.
+	DrawKey();
+	DrawSelectedSystem();
```
[the remark this answers is not in the document]
Comment (ChamEV) [reply source/MissionPanel.cpp]: It looks ok.  DrawMissionSystem draws the rings, DrawKey, DrawSelectedSystem only draw UI "windows", specifically the map key and the bar at the top center.

**Comment by Pointedstick** [comment]
> Still looks good. I don't really have any preference for one or the other though.
> Something I have noticed with both scrolling methods is that if I hold down my arrow key then I will select a different mission faster than the list can scroll, so the scrolling never catches up until I stop holding down the arrow key, at which point it quickly scrolls to the selected mission. Preferably, holding down the arrow key would have the scrolling keep up with the selected mission.
> 
> That makes me question though why the mission lists can't scroll the same way as the shipyard and outfitter scroll, referring to both the shipyard/outfitter while landed and the shipyard/outfitter on the map, which both will immediately snap to whatever is currently selected.

Yep, I agree.

Also in general for these scrollable views I'd really like scroll handles, but that's another subject. :)

**Comment by tehhowch** [comment]
Dropped the interpolated scroll for a "snap-to" style as used in ShipyardPanel, OutfitterPanel, etc.
Windows build here: https://ci.appveyor.com/project/tehhowch/endless-sky/builds/26238575/artifacts

**Pull request merged**